File: 0442_findDuplicates.py
class Solution:
    def findDuplicates(self, nums: List[int]) -> List[int]:
        for i in range(len(nums)):
            if nums[i] != i+1:
                while nums[i] != nums[nums[i]-1]:
                    a, b = nums[i], nums[nums[i]-1]
                    nums[nums[i]-1], nums[i] = a, b
        result = []
        for i in range(len(nums)):
            if nums[i] != i + 1:
                result.append(nums[i])
        return result

# Alternative solution

class Solution:
    def findDuplicates(self, nums):
        i = 0
        while i < len(nums):
            j = nums[i] - 1
            if nums[i] != nums[j]:
                nums[i], nums[j] = nums[j], nums[i]
            else:
                i += 1
        duplicates = []
        for i, num in enumerate(nums):
            if num != i + 1:
                duplicates.append(num)
        return duplicates

File: 0001-0500/0442_findDuplicates.java
class Solution {
    public List<Integer> findDuplicates(int[] nums) {
        List<Integer> duplicates = new ArrayList<>();
        int i = 0;
        int temp;

        while (i < nums.length) {
            while (nums[i] != i + 1 && nums[i] != nums[nums[i] - 1]) {
                temp = nums[i];
                nums[i] = nums[nums[i] - 1];
                nums[temp - 1] = temp;
            }
            i++;
        }

        for (int j = 0; j < nums.length; j++) {
            if (j != nums[j] - 1) {
                duplicates.add(nums[j]);
            }
        }

        return duplicates;
    }
}
